python/ray/data/datasource/dbapi2_datasource.py

```diff
@@ -0,0 +1,122 @@
+import math
+from contextlib import contextmanager
+from functools import lru_cache
+from typing import Any, Callable, Generator, Iterable, List, NewType, Optional
+
+import pyarrow as pa
+
+from ray.data._internal.arrow_block import ArrowRow
+from ray.data.block import Block, BlockAccessor, BlockMetadata
+from ray.data.datasource.datasource import Datasource, Reader, ReadTask, WriteResult
+
+# A Python DB API2-compliant `Connection` object.
+Connection = Any
+
+
+def cursor_to_block(cursor) -> Block:
+    rows = cursor.fetchall()
+    columns = [column_description[0] for column_description in cursor.description]
+    pylist = [{c: v for c, v in zip(columns, row)} for row in rows]
+    return pa.Table.from_pylist(pylist)
+
+
+class DBAPI2Datasource(Datasource[ArrowRow]):
+    def __init__(self, connection_factory: Callable[[], Connection]):
+        self.connection_factory = connection_factory
+
+    def create_reader(self, sql: str) -> "Reader":
+        return DBAPI2Reader(sql, self.connection_factory)
+
+
+@contextmanager
+def connect(
+    connection_factory: Callable[[], Connection]
+) -> Generator[Connection, None, None]:
+    connection = connection_factory()
+    try:
+        yield connection
+    except:
+        # `rollback` is optional since not all databases provide transaction support.
+        if hasattr(connection, "rollback"):
+            connection.rollback()
+    finally:
+        connection.close()
+
+
+class DBAPI2Reader(Reader):
+
+    NUM_SAMPLE_ROWS = 100
+
+    def __init__(self, sql: str, connection_factory: Callable[[], Connection]):
+        self.sql = sql
+        self.connection_factory = connection_factory
+
+    @property
+    @lru_cache
+    def sample_block(self) -> Block:
+        with connect(self.connection_factory) as connection:
+            cursor = connection.cursor()
+            cursor.execute(
+                f"SELECT * FROM ({self.sql}) as T LIMIT {self.NUM_SAMPLE_ROWS}"
+            )
+            return cursor_to_block(cursor)
+
+    @property
+    @lru_cache
+    def num_rows(self) -> int:
+        with connect(self.connection_factory) as connection:
+            cursor = connection.cursor()
+            cursor.execute(f"SELECT COUNT(*) FROM ({self.sql}) as T")
```

Comment:
This essentially executing the whole SQL query, and get number of rows of the result. So it might be quite inefficient - worst case, it doubles the cost of SQL query here, and here it's running on driver. 

Can we run it as a ray remote task w/ NodeAffinitySchedulingStrategy?

btw we should check if it's on Ray client, if it is, then we shouldn't run this on client side. Similar logic in https://github.com/ray-project/ray/blob/master/python/ray/data/read_api.py#L308-L317 .